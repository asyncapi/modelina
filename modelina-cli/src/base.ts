import { Command } from '@oclif/core';

export default abstract class extends Command {
  async catch(err: Error & { exitCode?: number; }): Promise<any> {
    try {
      return await super.catch(err);
    } catch (e: any) {
      if (e instanceof Error) {
        this.logToStderr(`${e.name}: ${e.message}`);
        process.exitCode = 1;
      }
    }
  }
}
