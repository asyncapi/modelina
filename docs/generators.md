# Generators

**Generators** are classes that are used to render models for a given language.

## Supported languages

- [JavaScript](../src/generators/javascript/JavaScriptGenerator.ts),
- [TypeScript](../src/generators/typescript/TypeScriptGenerator.ts),
- [Java](../src/generators/java/JavaGenerator.ts).
- [Go](../src/generators/go/GoGenerator.ts).
- [C#](../src/generators/csharp/CSharpGenerator.ts).

## Generator's options

Each generator extends default options for the generator. It means that the generator can also have additional options.

Options are passed as the first argument to the generator's constructor. Check the example:

```ts
const generator = new TypeScriptGenerator({ ...options });
```

Default options contain:

| Option | Type | Description | Default value |
|---|---|---|---|
| `indentation` | Object | Options for indentation. | - |
| `indentation.type` | String | Type of indentation. Its value can be either `SPACES` or `TABS`. | `SPACES` |
| `indentation.size` | String | Size of indentation. | 2 |
| `defaultPreset` | Object | Default preset for generator. For more information, read [customization](./customization.md) document. | _Implemented by generator_ |
| `presets` | Array | Array contains **presets**. For more information, read [customization](./customization.md) document. | `[]` |

Below is a list of additional options available for a given generator.

### [TypeScript](../src/generators/typescript/TypeScriptGenerator.ts)

| Option | Type | Description | Default value |
|---|---|---|---|
| `renderTypes` | Boolean | Render signature for types. | `true` |
| `modelType` | String | It indicates which model type should be rendered for the `object` type. Its value can be either `interface` or `class`. | `class` |
| `namingConvention` | Object | Options for naming conventions. | - |
| `namingConvention.type` | Function | A function that returns the format of the type. | _Returns pascal cased name, and ensures that reserved keywords are never rendered__ |
| `namingConvention.property` | Function | A function that returns the format of the property. | _Returns camel cased name, and ensures that names of properties does not clash against reserved keywords for TS, as well as JS to ensure painless transpilation_ |

### [Java](../src/generators/java/JavaGenerator.ts)

| Option | Type | Description | Default value |
|---|---|---|---|
| `collectionType` | String | It indicates with which signature should be rendered the `array` type. Its value can be either `List` (`List<{type}>`) or `Array` (`{type}[]`). | `List` |
| `namingConvention` | Object | Options for naming conventions. | - |
| `namingConvention.type` | Function | A function that returns the format of the type. | _Returns pascal cased name, and ensures that reserved keywords are never rendered__ |
| `namingConvention.property` | Function | A function that returns the format of the property. | _Returns camel cased name, and ensures that names of properties does not clash against reserved keywords_ |

### [JavaScript](../src/generators/javascript/JavaScriptGenerator.ts)

| Option | Type | Description | Default value |
|---|---|---|---|
| `namingConvention` | Object | Options for naming conventions. | - |
| `namingConvention.type` | Function | A function that returns the format of the type. | _Returns pascal cased name, and ensures that reserved keywords are never rendered_ |
| `namingConvention.property` | Function | A function that returns the format of the property. | _Returns camel cased name, and ensures that names of properties does not clash against reserved keywords_ |

### [Go](../src/generators/javascript/GoGenerator.ts)

| Option | Type | Description | Default value |
|---|---|---|---|
| `namingConvention` | Object | Options for naming conventions. | - |
| `namingConvention.type` | Function | A function that returns the format of the type. | _Returns pascal cased name_ |
| `namingConvention.field` | Function | A function that returns the format of the field. | _Returns pascal cased name_ |

### [C#](../src/generators/csharp/CSharpGenerator.ts)

| Option | Type | Description | Default value |
|---|---|---|---|
| `namingConvention` | Object | Options for naming conventions. | - |
| `namingConvention.type` | Function | A function that returns the format of the type. | _Returns pascal cased name, and ensures that reserved keywords are never rendered__ |
| `namingConvention.property` | Function | A function that returns the format of the property. | _Returns camel cased name, and ensures that names of properties does not clash against reserved keywords_ |

## Custom generator

The minimum set of required actions to create a new generator are:

- Source code must be included in [generators](../src/generators) folder.
- Must extend the abstract [`AbstractGenerator`](../src/generators/AbstractGenerator.ts) class,
- Must define [`Preset`](./customization.md)'s shape for the language,
- Must define language options by passing an interface describing additional options to the first generic argument of [`AbstractGenerator`](../src/generators/AbstractGenerator.ts). The interface must also be extended by `CommonGeneratorOptions` interface,
- Must define default options as static class's field, which must be extended by `defaultGeneratorOptions`,
- Default options must include `defaultPreset` property,
- Must implement `render` function,
- Must define **Renderers** classes for available model types in a given language. **Renderer** is an instance of the class with common helper functions to render appropriate model type and must be extended by [`AbstractRenderer`](../src/generators/AbstractRenderer.ts) class - [example](../src/generators/typescript/renderers/ClassRenderer.ts).

Check the [generator implementation](../src/generators/typescript/TypeScriptGenerator.ts) for `TypeScript` language to see how it should look like.

If you created a generator then you can contribute it to the AsyncAPI Model SDK and it will become the official supported generator.
