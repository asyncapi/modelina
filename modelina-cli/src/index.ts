export {run} from '@oclif/core'

export {generateModels, ModelinaArgs, ModelinaFlags, Languages} from './helpers/generate'