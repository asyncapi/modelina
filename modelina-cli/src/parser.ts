import { AvroSchemaParser } from '@asyncapi/avro-schema-parser';
import { OpenAPISchemaParser } from '@asyncapi/openapi-schema-parser';
import { Parser, convertToOldAPI } from '@asyncapi/parser/cjs';
import { RamlDTSchemaParser } from '@asyncapi/raml-dt-schema-parser';
import { Flags } from '@oclif/core';
import { ProtoBuffSchemaParser } from '@asyncapi/protobuf-schema-parser';
import { getDiagnosticSeverity } from '@stoplight/spectral-core';
import { OutputFormat } from '@stoplight/spectral-cli/dist/services/config';
import { html, json, junit, pretty, stylish, teamcity, text } from '@stoplight/spectral-formatters';

import type { Diagnostic } from '@asyncapi/parser/cjs';
import type Command from './base';
import type { Specification } from './models/SpecificationFile';

export type SeveritytKind = 'error' | 'warn' | 'info' | 'hint';

export { convertToOldAPI };

const parser = new Parser({
  __unstable: {
    resolver: {
      cache: false,
    }
  }
});

parser.registerSchemaParser(AvroSchemaParser());
parser.registerSchemaParser(OpenAPISchemaParser());
parser.registerSchemaParser(RamlDTSchemaParser());
parser.registerSchemaParser(ProtoBuffSchemaParser());

export interface ValidationFlagsOptions {
  logDiagnostics?: boolean;
}

export function validationFlags({ logDiagnostics = true }: ValidationFlagsOptions = {}) {
  return {
    'log-diagnostics': Flags.boolean({
      description: 'log validation diagnostics or not',
      default: logDiagnostics,
      allowNo: true,
    }),
    'diagnostics-format': Flags.enum({
      description: 'format to use for validation diagnostics',
      options: Object.values(OutputFormat),
      default: OutputFormat.STYLISH,
    }),
    'fail-severity': Flags.enum<SeveritytKind>({
      description: 'diagnostics of this level or above will trigger a failure exit code',
      options: ['error', 'warn', 'info', 'hint'],
      default: 'error',
    }),
  };
}

interface ValidateOptions {
  'log-diagnostics'?: boolean;
  'diagnostics-format'?: `${OutputFormat}`;
  'fail-severity'?: SeveritytKind;
}

export async function validate(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const diagnostics = await parser.validate(specFile.text(), { source: specFile.getSource() });
  return logDiagnostics(diagnostics, command, specFile, options);
}

export async function parse(command: Command, specFile: Specification, options: ValidateOptions = {}) {
  const { document, diagnostics } = await parser.parse(specFile.text(), { source: specFile.getSource() });
  const status = logDiagnostics(diagnostics, command, specFile, options);
  return { document, diagnostics, status };
}

function logDiagnostics(diagnostics: Diagnostic[], command: Command, specFile: Specification, options: ValidateOptions = {}): 'valid' | 'invalid' {
  const logDiagnostics = options['log-diagnostics'];
  const failSeverity = options['fail-severity'] ?? 'error';
  const diagnosticsFormat = options['diagnostics-format'] ?? 'stylish';

  const sourceString = specFile.toSourceString();
  if (diagnostics.length) {
    if (hasFailSeverity(diagnostics, failSeverity)) {
      if (logDiagnostics) {
        command.logToStderr(`\n${sourceString} and/or referenced documents have governance issues.`);
        command.logToStderr(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
      }
      return 'invalid';
    }

    if (logDiagnostics) {
      command.log(`\n${sourceString} is valid but has (itself and/or referenced documents) governance issues.`);
      command.log(formatOutput(diagnostics, diagnosticsFormat, failSeverity));
    }
  } else if (logDiagnostics) {
    command.log(`\n${sourceString} is valid! ${sourceString} and referenced documents don't have governance issues.`);
  }

  return 'valid';
}

export function formatOutput(diagnostics: Diagnostic[], format: `${OutputFormat}`, failSeverity: SeveritytKind) {
  const options = { failSeverity: getDiagnosticSeverity(failSeverity) };
  switch (format) {
  case 'stylish': return stylish(diagnostics, options);
  case 'json': return json(diagnostics, options);
  case 'junit': return junit(diagnostics, options);
  case 'html': return html(diagnostics, options);
  case 'text': return text(diagnostics, options);
  case 'teamcity': return teamcity(diagnostics, options);
  case 'pretty': return pretty(diagnostics, options);
  default: return stylish(diagnostics, options);
  }
}

function hasFailSeverity(diagnostics: Diagnostic[], failSeverity: SeveritytKind) {
  const diagnosticSeverity = getDiagnosticSeverity(failSeverity);
  return diagnostics.some(diagnostic => diagnostic.severity <= diagnosticSeverity);
}
