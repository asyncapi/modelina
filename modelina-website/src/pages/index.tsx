
import Button from '@/components/buttons/Button';
import GithubButton from '@/components/buttons/GithubButton';
import CodeBlock from '@/components/CodeBlock';
import IconRocket from '@/components/icons/Rocket';
import GenericLayout from '@/components/layouts/GenericLayout';
import Heading from '@/components/typography/Heading';
import Paragraph from '@/components/typography/Paragraph';
import Image from 'next/image';

export default function Index() {
  const description =
    'Sometimes you just want to generate data models for your payload. Modelina is a library for generating data models based on inputs such as AsyncAPI, OpenAPI, or JSON Schema documents.';
  const image = '/img/social/modelina-card.jpg';

  return (
    <GenericLayout
      title="Modelina"
      description={description}
      image={image}
      wide
    >
      <div className="py-16 overflow-hidden lg:py-24">
        <div className="relative text-center">
          <Heading level="h1" typeStyle="heading-lg">
            Modelina
          </Heading>
          <Paragraph className="mt-4 max-w-3xl mx-auto">
            {description}
          </Paragraph>
        </div>

        <div className="relative mt-12 lg:mt-20 lg:grid lg:grid-cols-2 lg:gap-8 lg:items-center">
          <div className="relative mb-8 lg:mt-8">
            <Heading level="h4" typeStyle="heading-md-semibold">
              AsyncAPI Cli
            </Heading>
            <Paragraph className="mt-3 lg:pr-4">
              Get started immediately with Modelina through the AsyncAPI Cli.
            </Paragraph>
            <div className="mt-8">
              <CodeBlock
                language="generator-cli"
                showLineNumbers={false}
                className="mt-8"
                textSizeClassName="text-sm"
              >
                asyncapi generate models typescript ./asyncapi.json
              </CodeBlock>
              <div className="mt-8">
                <GithubButton
                  className="block mt-2 md:mt-0 md:inline-block w-full sm:w-auto mt-8"
                  href="https://github.com/asyncapi/cli"
                />
              </div>
            </div>
          </div>
          <div className="relative lg:mt-8 h-full">
            <Image
              src={'/img/card/cli-card.jpg'}
              fill
              sizes='100vw'
              alt={'CLI card'}
            />
          </div>
        </div>

        <div className="relative mt-12 lg:mt-20 lg:grid lg:grid-cols-2 lg:gap-8 lg:items-center">
          <div className="relative mb-8 lg:mt-8">
            <Heading level="h4" typeStyle="heading-md-semibold">
              Installation
            </Heading>
            <Paragraph className="mt-3 lg:pr-4">
              Install Modelina as a library to take full control.
            </Paragraph>
            <div className="mt-8">
              <CodeBlock
                language="bash"
                showLineNumbers={false}
                className="mt-8"
                textSizeClassName="text-sm"
              >
                npm install @asyncapi/modelina
              </CodeBlock>
              <div className="mt-8">
                <Button
                  className="hidden mt-2 md:mt-0 lg:inline-block md:ml-2"
                  text="Try The Playground"
                  href="/playground"
                  icon={<IconRocket className="inline-block -mt-1 w-6 h-6" />}
                />
              </div>
            </div>
          </div>
          <div className = "relative lg:mt-8 h-full">
            <Image 
            src = {'/img/card/modelina-card.jpg'}
            fill 
            sizes='100vw'
            alt = {'Modelina card'}
          />
          </div>
        </div>
        
        <div className="relative text-center mt-12">
          <Heading level="h1" typeStyle="heading-lg">
            Usage
          </Heading>
          <Paragraph className="mt-4 max-w-3xl mx-auto">
            Modelina can be used in a wide range of scenarios, and is build to
            be apart of the core toolbox you can customize to your hearts
            desire.{' '}
            <b>Here is just a fraction of what you can do with Modelina.</b>
          </Paragraph>
        </div>

        <div className="relative lg:mt-20 lg:items-center">
          <table className="min-w-full relative lg:mt-8">
            <tbody>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Modelina lets you generate data models from many types of{' '}
                    <a href="#inputs">inputs</a>
                  </b>
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const asyncapi = ...
const jsonschema = ...
const openapi = ...
const metamodel = ...
...
const models = await generator.generate(
  asyncapi | jsonschema | openapi | metamodel
);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Use the same inputs across a range of different{' '}
                    <a href="#outputs">generators</a>
                  </b>
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator();
const generator = new CsharpGenerator();
const generator = new JavaGenerator();
const generator = new RustGenerator();
...
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>Easily let you interact with the generated models.</b>-
                  Want to show the generated models on a website? Sure! - Want
                  to generate the models into files? Sure! - Want to combine all
                  the models into one single file? Sure! Whatever interaction
                  you need, you can create.
                </td>
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const models = await generator.generate(input);
for (const model in models) { 
  const generatedCode = generatedModel.result;
  const dependencies = generatedModel.dependencies;
  const modeltype = generatedModel.model.type;
  const modelName = generatedModel.modelName;
  ...
}`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Easily modify how models are{' '}
                    <a href="./docs/constraints.md">constrained</a> into the
                    output
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator({
  constraints: {
    modelName: ({modelName}) => {
      // Implement your own constraining logic
      return modelName;
    }
  }
});`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Seamlessly layer additional or replacement code{' '}
                    <a href="./docs/presets.md">
                      on top of each other to customize the models
                    </a>{' '}
                    to your use-case
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const generator = new TypeScriptGenerator({
  presets: [
    {
      class: {
        additionalContent({ content }) {
          return \`\${content}
public myCustomFunction(): string {
  return 'A custom function for each class';
}\`;
        },
      }
    }
  ]
});
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
              <tr className="border-b">
                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <b>
                    Seamlessly lets you{' '}
                    <a href="./docs/presets.md">
                      combine multiple layers of additional or replacement code
                    </a>
                  </b>
                </td>

                <td className="text-sm text-gray-900 font-light px-6 py-4">
                  <CodeBlock
                    language="typescript"
                    showLineNumbers={false}
                    className="mt-8"
                    textSizeClassName="text-sm"
                  >
                    {`const myCustomFunction1 = {
  class: {
    additionalContent({ content }) {
      return \`\${content}
public myCustomFunction(): string {
return 'A custom function for each class';
}\`;
    },
  }
};
const myCustomFunction2 = {...};
const generator = new TypeScriptGenerator({
  presets: [
    myCustomFunction1,
    myCustomFunction2
  ]
});
const models = await generator.generate(input);`}
                  </CodeBlock>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </GenericLayout>
  );
}
