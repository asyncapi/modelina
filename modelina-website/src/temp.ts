import { InputMetaModel, ProcessorOptions, AbstractInputProcessor } from '@asyncapi/modelina';

/** Class for processing Typescript code inputs to Common module*/
export interface TypeScriptInputProcessorOptions {
  uniqueNames?: boolean;
  required?: boolean;
}
export class TypeScriptInputProcessor extends AbstractInputProcessor {
  shouldProcess(input: any): boolean {
    return false;
  }

  process(input: any, options?: ProcessorOptions): Promise<InputMetaModel> {
    return Promise.reject();
  }
}
