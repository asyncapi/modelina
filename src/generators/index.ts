export * from './AbstractGenerator';
export * from './AbstractRenderer';
export * from './java';
export * from './dart';
export * from './csharp';
export * from './javascript';
export * from './typescript';
export * from './python';
export * from './go';
export * from './rust';
export * from './kotlin';
export * from './cplusplus';
export * from './AbstractFileGenerator';
