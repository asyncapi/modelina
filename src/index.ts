export * from './generators';
export * from './helpers';
export * from './models';
export * from './processors';
export type { ModelLoggingInterface } from './utils';
export { Logger } from './utils';
