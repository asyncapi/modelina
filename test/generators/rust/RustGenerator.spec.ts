import { defaultRustRenderCompleteModelOptions, RustGenerator, RustRenderCompleteModelOptions } from '../../../src/generators';

describe('RustGenerator', () => {
  let generator: RustGenerator;
  beforeEach(() => {
    generator = new RustGenerator();
  });

  describe('Enum', () => {
    test('should not render reserved keyword', async () => {
      const doc = {
        $id: 'Address',
        type: 'object',
        properties: {
          enum: { type: 'string' },
          reservedEnum: { type: 'string' }
        },
        additionalProperties: false
      };

      const models = await generator.generate(doc);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });
    test('should render `enum` type (integer type)', async () => {
      const doc = {
        $id: 'Numbers',
        type: 'integer',
        enum: [0, 1, 2, 3],
      };

      const models = await generator.generate(doc);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });

    test('should render `enum` with mixed types (union type)', async () => {
      const doc = {
        $id: 'Things_123',
        enum: ['Texas', 1, '1', false, { test: 'test' }],
      };
      const models = await generator.generate(doc);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });

    test('should work custom preset for `enum` type', async () => {
      const doc = {
        $id: 'CustomEnum',
        type: 'string',
        enum: ['Texas', 'Alabama', 'California'],
      };

      generator = new RustGenerator({
        presets: [
          {
            enum: {
              self({ content }) {
                return content;
              },
            }
          }
        ]
      });

      const models = await generator.generate(doc);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });

    test('should render implement Default for `enum (default provided)', async () => {
      // if Default is not specified, the first item in enum is implemented instead
      const doc = {
        $id: 'CustomEnum',
        type: 'string',
        default: 'Texas',
        enum: ['Texas', 'Alabama', 'California'],
      };

      const options = { ...defaultRustRenderCompleteModelOptions, implementDefault: true, packageName: 'test' } as RustRenderCompleteModelOptions;

      const models = await generator.generateCompleteModels(doc, options);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });

    test('should render enums with translated special characters', async () => {
      const doc = {
        $id: 'States',
        enum: ['test+', '$test', 'test-', 'test?!', '*test']
      };
      const models = await generator.generate(doc);
      expect(models).toHaveLength(1);
      expect(models[0].result).toMatchSnapshot();
    });
  });
  describe('Struct & Complete Models', () => {
    const doc = {
      $id: '_address',
      type: 'object',
      properties: {
        street_name: { type: 'string' },
        city: { type: 'string', description: 'City description' },
        state: { type: 'string' },
        house_number: { type: 'number' },
        marriage: { type: 'boolean', description: 'Status if marriage live in given house' },
        members: { oneOf: [{ type: 'string' }, { type: 'number' }, { type: 'boolean' }], },
        tuple_type: { type: 'array', items: [{ type: 'string' }, { type: 'number' }], additionalItems: false },
        array_type: { type: 'array', items: { type: 'string' }, additionalItems: false },
        // not yet implemented
        // tuple_type_with_untyped_additional_items: { type: 'array', items: [{ type: 'string' }, { type: 'number' }], additionalItems: true },
        // tuple_type_with_typed_additional_items: { type: 'array', items: [{ type: 'string' }, { type: 'number' }], additionalItems: { type: 'string' } },
        // array_type_with_typed_additional_items: { type: 'array', items: { type: 'string' }, additionalItems: { type: 'string' } },
        // array_type_with_any_additional_items: { type: 'array', items: { type: 'string' }, additionalItems: true },

      },
      required: ['street_name', 'city', 'state', 'house_number', 'array_type'],
      additionalProperties: {
        type: 'string'
      },
      // not yet implemented
      // patternProperties: {
      //     '^S(.?*)test&': {
      //         type: 'string'
      //     }
      // },
    };

    test('should render `struct` type ', async () => {
      const models = await generator.generate(doc);
      expect(models).toHaveLength(3); // Adress, TupleType, and Member models should be generated
      expect(models[0].result).toMatchSnapshot(); // Address model
      expect(models[1].result).toMatchSnapshot(); // Member model
      expect(models[2].result).toMatchSnapshot(); // TupleType model
    });

    test('Should render complete models', async () => {
      const options = { ...defaultRustRenderCompleteModelOptions, implementDefault: true, implementNew: true, supportFiles: true } as RustRenderCompleteModelOptions;

      const models = await generator.generateCompleteModels(doc, options);
      expect(models).toHaveLength(3);
      expect(models[0].result).toMatchSnapshot(); // Address model
      expect(models[1].result).toMatchSnapshot(); // Member model
      expect(models[2].result).toMatchSnapshot(); // TupleType model
    });
  });

  describe('Packaging', () => {
    test('Should render supporting files', async () => {
      const doc = {
        $id: 'Numbers',
        type: 'integer',
        enum: [0, 1, 2, 3],
      };
      const options = { ...defaultRustRenderCompleteModelOptions, implementDefault: true, implementNew: true, supportFiles: true } as RustRenderCompleteModelOptions;
      const output = await generator.generateCompleteSupport(doc, options);
      expect(output).toHaveLength(2);
      expect(output[0].result).toMatchSnapshot(); // Cargo.toml
      expect(output[1].result).toMatchSnapshot(); // lib.rs
    });
  });
});
