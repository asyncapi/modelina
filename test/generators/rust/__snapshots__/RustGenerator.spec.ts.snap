// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RustGenerator Enum should not render reserved keyword 1`] = `
"// Address represents a Address model.
#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub struct Address {
    #[serde(rename=\\"reservedEnum\\", skip_serializing_if = \\"Option::is_none\\")]
    reserved_enum: Option<String>,
}
"
`;

exports[`RustGenerator Enum should render \`enum\` type (integer type) 1`] = `
"// Numbers represents a Numbers model.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub enum Numbers {
    #[serde(rename=\\"0\\")]
    Number_0,
    #[serde(rename=\\"1\\")]
    Number_1,
    #[serde(rename=\\"2\\")]
    Number_2,
    #[serde(rename=\\"3\\")]
    Number_3,
}
"
`;

exports[`RustGenerator Enum should render \`enum\` with mixed types (union type) 1`] = `
"// Things represents a Things model.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub enum Things {
    #[serde(rename=\\"Texas\\")]
    Texas,
    #[serde(rename=\\"1\\")]
    Number_1,
    #[serde(rename=\\"1\\")]
    ReservedNumber_1,
    #[serde(rename=\\"false\\")]
    ReservedFalse,
    #[serde(flatten)]
    CurlyleftQuotationTestQuotationColonQuotationTestQuotationCurlyright(HashMap<String, serde_json::Value>),
}
"
`;

exports[`RustGenerator Enum should render enums with translated special characters 1`] = `
"// States represents a States model.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub enum States {
    #[serde(rename=\\"test+\\")]
    TestPlus,
    #[serde(rename=\\"$test\\")]
    DollarTest,
    #[serde(rename=\\"test-\\")]
    TestMinus,
    #[serde(rename=\\"test?!\\")]
    TestQuestionExclamation,
    #[serde(rename=\\"*test\\")]
    AsteriskTest,
}
"
`;

exports[`RustGenerator Enum should render implement Default for \`enum (default provided) 1`] = `
"// CustomEnum represents a CustomEnum model.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub enum CustomEnum {
    #[serde(rename=\\"Texas\\")]
    Texas,
    #[serde(rename=\\"Alabama\\")]
    Alabama,
    #[serde(rename=\\"California\\")]
    California,
}
"
`;

exports[`RustGenerator Enum should work custom preset for \`enum\` type 1`] = `
"// CustomEnum represents a CustomEnum model.
#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
pub enum CustomEnum {
    #[serde(rename=\\"Texas\\")]
    Texas,
    #[serde(rename=\\"Alabama\\")]
    Alabama,
    #[serde(rename=\\"California\\")]
    California,
}
"
`;

exports[`RustGenerator Packaging Should render supporting files 1`] = `
"[package]
name = \\"asyncapi-rs-models\\"
version = \\"0.0.0\\"
authors = [\\"AsyncAPI Rust Champions\\"]
homepage = \\"https://www.asyncapi.com/tools/modelina\\"
repository = \\"https://github.com/asyncapi/modelina\\"
license = \\"Apache-2.0\\"
description = \\"Rust models generated by AsyncAPI Modelina\\"
edition = \\"2018\\"

[dependencies]
serde = { version = \\"1\\", features = [\\"derive\\"] }
serde_json = { version=\\"1\\", optional = true }

[dev-dependencies]

[features]
default = [\\"json\\"]
json = [\\"dep:serde_json\\"]"
`;

exports[`RustGenerator Packaging Should render supporting files 2`] = `
"#[macro_use]
extern crate serde;
extern crate serde_json;

pub mod numbers;
pub use self::numbers::*;"
`;

exports[`RustGenerator Struct & Complete Models Should render complete models 1`] = `
"// Address represents a Address model.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct Address {
    #[serde(rename=\\"street_name\\")]
    street_name: String,
    #[serde(rename=\\"city\\")]
    city: String,
    #[serde(rename=\\"state\\")]
    state: String,
    #[serde(rename=\\"house_number\\")]
    house_number: f64,
    #[serde(rename=\\"marriage\\", skip_serializing_if = \\"Option::is_none\\")]
    marriage: Option<bool>,
    #[serde(rename=\\"members\\", skip_serializing_if = \\"Option::is_none\\")]
    members: Option<Box<crate::Members>>,
    #[serde(rename=\\"tuple_type\\", skip_serializing_if = \\"Option::is_none\\")]
    tuple_type: Option<Box<crate::TupleType>>,
    #[serde(rename=\\"array_type\\")]
    array_type: Vec<String>,
    #[serde(rename=\\"additionalProperties\\", skip_serializing_if = \\"Option::is_none\\")]
    additional_properties: Option<std::collections::HashMap<String, String>>,
}
"
`;

exports[`RustGenerator Struct & Complete Models Should render complete models 2`] = `
"// Members represents a union of types: String, f64, bool
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub enum Members {
    #[serde(rename=\\"Members0\\")]
    Members0(String),
    #[serde(rename=\\"Members1\\")]
    Members1(f64),
    #[serde(rename=\\"Members2\\")]
    Members2(bool),
}

"
`;

exports[`RustGenerator Struct & Complete Models Should render complete models 3`] = `
"// TupleType represents a TupleType model.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct TupleType(String, f64);

"
`;

exports[`RustGenerator Struct & Complete Models should render \`struct\` type  1`] = `
"// Address represents a Address model.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct Address {
    #[serde(rename=\\"street_name\\")]
    street_name: String,
    #[serde(rename=\\"city\\")]
    city: String,
    #[serde(rename=\\"state\\")]
    state: String,
    #[serde(rename=\\"house_number\\")]
    house_number: f64,
    #[serde(rename=\\"marriage\\", skip_serializing_if = \\"Option::is_none\\")]
    marriage: Option<bool>,
    #[serde(rename=\\"members\\", skip_serializing_if = \\"Option::is_none\\")]
    members: Option<Box<crate::Members>>,
    #[serde(rename=\\"tuple_type\\", skip_serializing_if = \\"Option::is_none\\")]
    tuple_type: Option<Box<crate::TupleType>>,
    #[serde(rename=\\"array_type\\")]
    array_type: Vec<String>,
    #[serde(rename=\\"additionalProperties\\", skip_serializing_if = \\"Option::is_none\\")]
    additional_properties: Option<std::collections::HashMap<String, String>>,
}
"
`;

exports[`RustGenerator Struct & Complete Models should render \`struct\` type  2`] = `
"// Members represents a union of types: String, f64, bool
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub enum Members {
    #[serde(rename=\\"Members0\\")]
    Members0(String),
    #[serde(rename=\\"Members1\\")]
    Members1(f64),
    #[serde(rename=\\"Members2\\")]
    Members2(bool),
}

"
`;

exports[`RustGenerator Struct & Complete Models should render \`struct\` type  3`] = `
"// TupleType represents a TupleType model.
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct TupleType(String, f64);

"
`;
